### PR TITLE
Switch from macro based StrongTypedef to template version

### PR DESCRIPTION
**Pull request by TheMarex**
# Issue

This enables us to use lazy instantiation of member functions. Namely we an now have a `operator|` that does not cause compile errors for `From = double` (only when one actually tries to use it).

## Tasklist
 - [x] review
 - [x] adjust for comments

Needed for #3952

### CODE REVIEW

File: include/util/alias.hpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#ifndef ALIAS_HPP
+#define ALIAS_HPP
```
Comment (daniel-j-h) [inline include/util/alias.hpp]: OSRM_

File: include/util/alias.hpp [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    return stream << inst.__value;
+}
+
+template <typename T>
+struct is_valid_alias
+    : std::integral_constant<bool,
+                             std::is_trivial<T>::value && std::is_standard_layout<T>::value &&
+                                 std::is_pod<T>::value>
+{
+};
```
Comment (daniel-j-h) [inline include/util/alias.hpp]: Hmm why not e.g. (not tested)

    template <typename T, typename Alias>
    struct is_alias_for : is_same<typename Alias::From, T> { };

    template <typename T, typename Alias>
    void assert_alias_for( static_assert(is_alias_for<T, Alias>, ""); );

Then we can check e.g.

    assert_alias_for<int32_t, OsmId>();

File: include/util/alias.hpp [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+#include <type_traits>
+
+namespace osrm
+{
+
+template <typename From, typename Tag> struct Alias;
+template <typename From, typename Tag>
+inline std::ostream &operator<<(std::ostream &stream, const Alias<From, Tag> &inst);
+
+template <typename From, typename Tag> struct Alias final
```
Comment (daniel-j-h) [inline include/util/alias.hpp]: If you make `Alias` a bit smarter and special case for primitive types vs. non-primitive types, you can inherit from non-primitive types and :sparkles: we have proper newtype semantics e.g. for vectors:

https://github.com/mapbox/node-or-tools/blob/master/src/types.h#L13-L23

compiler will complain if you mix things up, conversion is still possible to its base type

File: include/engine/routing_algorithms/routing_base.hpp [hunk 4]
```diff
@@ -320,17 +320,19 @@ double getPathDistance(const datafacade::ContiguousInternalMemoryDataFacade<Algo
...
 
         const double current_lat =
-            static_cast<double>(toFloating(current_coordinate.lat)) * DEGREE_TO_RAD;
+            static_cast<double>(util::toFloating(current_coordinate.lat)) * DEGREE_TO_RAD;
         const double current_lon =
-            static_cast<double>(toFloating(current_coordinate.lon)) * DEGREE_TO_RAD;
+            static_cast<double>(util::toFloating(current_coordinate.lon)) * DEGREE_TO_RAD;
         const double current_cos = std::cos(current_lat);
 
         const double sin_dlon = std::sin((prev_lon - current_lon) / 2.0);
```
Comment (oxidase) [inline include/engine/routing_algorithms/routing_base.hpp]: if types now have proper `operator-` and `operator/(double)` then here degrees be subtracted first and divided by 2 and than converted to double radians

File: include/util/alias.hpp [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    }
+    Alias operator-(const Alias rhs_) const
+    {
+        return Alias{__value - static_cast<const From>(rhs_)};
+    }
+    Alias operator*(const Alias rhs_) const
+    {
+        return Alias{__value * static_cast<const From>(rhs_)};
+    }
+    Alias operator/(const Alias rhs_) const
```
Comment (oxidase) [inline include/util/alias.hpp]: here better to add also `Alias operator/(const double denom) const` for scaling

File: include/util/alias.hpp [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+inline std::ostream &operator<<(std::ostream &stream, const Alias<From, Tag> &inst)
+{
+    return stream << inst.__value;
+}
+
+template <typename T>
+struct is_valid_alias
+    : std::integral_constant<bool,
+                             std::is_trivial<T>::value && std::is_standard_layout<T>::value &&
+                                 std::is_pod<T>::value>
```
Comment (oxidase) [inline include/util/alias.hpp]: i guess `is_pod` in c++11 terms is enough http://en.cppreference.com/w/cpp/types/is_pod

File: include/util/alias.hpp (hunk 2 above)
Comment (TheMarex) [inline include/util/alias.hpp]: The trait is more supposed to check if it really is just a type alias and will be optimized away. Will just check `is_pod` instead.

**Comment by TheMarex** [comment]
@danpat checked the assembly, seems to boil down to virtually the same instructions in both cases.

**Pull request merged**